common/faults/faults.c

```diff
@@ -0,0 +1,461 @@
+						critCount--;
+						break;
+				}
+			}
+			break;
+		}
+}
+
+//BEGIN AUTO RECIEVE FUNCTIONS
+void fault_sync_main_module_CALLBACK(CanParsedData_t *msg_header_a) {
+	fault_message_t recievedMessage = {msg_header_a->fault_sync_main_module.latched, msg_header_a->fault_sync_main_module.idx};
+	fault_message_t *currMessage = &messageArray[GET_IDX(recievedMessage.f_ID)];
+	handleCallbacks(recievedMessage, currMessage);
+}
+void fault_sync_driveline_CALLBACK(CanParsedData_t *msg_header_a) {
+	fault_message_t recievedMessage = {msg_header_a->fault_sync_driveline.latched, msg_header_a->fault_sync_driveline.idx};
+	fault_message_t *currMessage = &messageArray[GET_IDX(recievedMessage.f_ID)];
+	handleCallbacks(recievedMessage, currMessage);
+}
+void fault_sync_dashboard_CALLBACK(CanParsedData_t *msg_header_a) {
+	fault_message_t recievedMessage = {msg_header_a->fault_sync_dashboard.latched, msg_header_a->fault_sync_dashboard.idx};
+	fault_message_t *currMessage = &messageArray[GET_IDX(recievedMessage.f_ID)];
+	handleCallbacks(recievedMessage, currMessage);
+}
+void fault_sync_precharge_CALLBACK(CanParsedData_t *msg_header_a) {
+	fault_message_t recievedMessage = {msg_header_a->fault_sync_precharge.latched, msg_header_a->fault_sync_precharge.idx};
+	fault_message_t *currMessage = &messageArray[GET_IDX(recievedMessage.f_ID)];
+	handleCallbacks(recievedMessage, currMessage);
+}
+void fault_sync_torque_vector_CALLBACK(CanParsedData_t *msg_header_a) {
+	fault_message_t recievedMessage = {msg_header_a->fault_sync_torque_vector.latched, msg_header_a->fault_sync_torque_vector.idx};
+	fault_message_t *currMessage = &messageArray[GET_IDX(recievedMessage.f_ID)];
+	handleCallbacks(recievedMessage, currMessage);
+}
+void fault_sync_l4_testing_CALLBACK(CanParsedData_t *msg_header_a) {
+	fault_message_t recievedMessage = {msg_header_a->fault_sync_l4_testing.latched, msg_header_a->fault_sync_l4_testing.idx};
+	fault_message_t *currMessage = &messageArray[GET_IDX(recievedMessage.f_ID)];
+	handleCallbacks(recievedMessage, currMessage);
+}
+//END AUTO RECIEVE FUNCTIONS
+
+//Force faults from daq
+void set_fault_CALLBACK(CanParsedData_t *msg_header_a) {
+    if (GET_OWNER(msg_header_a->set_fault.id) == currentMCU) {
+        forceFault(msg_header_a->set_fault.id, msg_header_a->set_fault.value);
+    }
+}
+
+//Return control back to this mcu from daq
+void return_fault_control_CALLBACK(CanParsedData_t *msg_header_a) {
+    if (GET_OWNER(msg_header_a->set_fault.id) == currentMCU) {
+        unForce(msg_header_a->set_fault.id);
+    }
+}
+
+
+//Updates faults owned by current mcu
+void updateFaults() {
+    uint16_t idx = ownedidx;
+    fault_attributes_t *fault;
+    do {
+        fault = &faultArray[idx];
+        //Fault is showing up as latched
+        if (((fault->tempLatch) && !(fault->message->latched))) {
+            //Has latching period ended
+            fault->message->latched = (fault->time_since_latch++ >= faultLatchTime[idx]) ? true : false;
```

Comment:
If the fault becomes latched, I think you need to reset your time_since_latch so that you can count the unlatch time

Reply:
I can. In all cases where the fault isn't changing state, the latch is set to 0. So as long as it doesn't go from latched to unlatching IMMEDIATELY, it would be fine. However this is an edge case, and I've resolved it for both latching and unlatching